caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h

```diff
@@ -26,6 +26,87 @@ template <typename T, typename U>
 std::vector<Subgraph<T, U>> tarjans(Graph<T, U> *g);
 #include "nomnigraph/Graph/TarjansImpl.h"
 
+
+/// \brief A binary graph matching algorithm based on Kahn's algorithm.
+template <typename T, typename U, typename F>
+std::vector<Subgraph<T, U>> binaryMatch(Graph<T, U> *g, F condition) {
+  using G = Graph<T, U>;
+  auto edges = g->getMutableEdges();
+  std::unordered_set<typename G::EdgeRef> edgeSet(edges.begin(), edges.end());
+  // Topologically sorted nodes.
+  std::vector<Subgraph<T,U>> sortedNodes;
+  sortedNodes.emplace_back();
+
+  // Find the initial frontier.
+  std::vector<typename G::NodeRef> frontier;
+  std::vector<typename G::NodeRef> nextFrontier;
+
+  for (auto n : g->getMutableNodes()) {
+    if (n->getInEdges().size() == 0 && condition(n)) {
+      frontier.emplace_back(n);
+    }
+  }
+
+  auto stillHasInEdge = [&](typename G::NodeRef m) {
```

Comment:
This is not very important but using edgeset is not very efficient. I like the way you did for Tarjan's, where you attach an annotation property with the graph node.